File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/3010_divide-an-array-into-subarrays-with-minimum-cost-i/solutions/java/solution.java
class Solution {
    public int minimumCost(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n][4];
        final int INF = Integer.MAX_VALUE / 2; // Use a large value to represent infeasible splits

        // Initialize DP table
        for (int i = 0; i < n; ++i) {
            for (int k = 1; k <= 3; ++k) {
                dp[i][k] = INF;
            }
        }

        // Base case: the first element starts the first group
        dp[0][1] = nums[0];

        // Fill the DP table
        for (int i = 1; i < n; ++i) {
            for (int k = 1; k <= 3; ++k) {
                // Continue the current group
                if (k > 1 && i >= k) {
                    dp[i][k] = Math.min(dp[i][k], dp[i-1][k] + nums[i]);
                }
                // Start a new group
                if (k > 1 && dp[i-1][k-1] != INF) {
                    dp[i][k] = Math.min(dp[i][k], dp[i-1][k-1] + nums[i]);
                }
            }
        }

        // The answer is at the last element with exactly 3 groups formed
        return dp[n-1][3];
    }
}

File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/3010_divide-an-array-into-subarrays-with-minimum-cost-i/solutions/cpp/solution.cpp
#include <vector>
#include <algorithm>
#include <climits>

class Solution {
public:
    int minimumCost(std::vector<int>& nums) {
        int n = nums.size();
        std::vector<std::vector<int>> dp(n, std::vector<int>(4, INT_MAX));
        
        // Initialize the dp array for 1 subarray from each point i to n
        for (int i = 0; i < n; ++i) {
            dp[i][1] = nums[i];
        }

        // Fill dp for j = 2 to 3 subarrays
        for (int j = 2; j <= 3; ++j) {
            for (int i = 0; i <= n - j; ++i) {
                for (int k = i + 1; k <= n - j + 1; ++k) {
                    dp[i][j] = std::min(dp[i][j], nums[i] + dp[k][j - 1]);
                }
            }
        }

        // Answer is the minimum cost to split the whole array into 3 subarrays
        return dp[0][3];
    }
};
